### ISSUE TITLE
specification for minContains, maxContains should reference the applicator $vocabulary

### ISSUE THREAD

**Issue by karenetheridge** [json-schema-org/json-schema-spec#976]
contains, minContains and maxContains are tricky because they are spread across different vocabularies.

The specification for minContains and maxContains says:

> If "contains" is not present within the same schema object, then this keyword has no effect. 

It should also indicate that the keywords have no effect if the metaschema disables the applicator vocabulary in the $vocabulary keyword. This would be a strange thing to do, but it is legal, so we should say so. (And we need a test for this in the test suite, too, when we get to writing tests for custom metaschemas.) :)

**Comment by handrews** [comment]
@karenetheridge how is disabling the applicator vocabulary distinct from the keyword not being present?

We have avoided requiring implementations to understand which vocabulary provides a given keyword.  `minContains` and `maxContains` depend on the `contains` keyword.  If you swap in a different vocabulary that also uses `contains` but means something different, you get whatever bizarre behavior you get.  Just as if you tried to use two different vocabularies offering incompatible definitions for `contains`.

If we get to a machine-readable vocabulary/keyword specification then it might be possible to do something about this.  However, I am very strongly against cramming special-case vocabulary logic in here.

> if the metaschema disables the applicator vocabulary in the $vocabulary keyword.

I also want to push back on this terminology.  Metaschemas don't disable vocabularies.  They either have them or they don't.  They don't have them but turn them off.

**Comment by Relequestual** [comment]
There's potentially a whole HOST of things that could go wrong or wouldn't work if the applicator vocabulary was not understood. This issue highlights just one of them.
I don't want to spend any energy on that till we have a concreet usecase or a real problem we cannot clearly solve.
I want to avoid hypertheticals on this.

**Comment by karenetheridge** [comment]
> minContains and maxContains depend on the contains keyword. If you swap in a different vocabulary that also uses contains but means something different, you get whatever bizarre behavior you get.

Let me rephrase: if a schema contains the keyword `contains`, but the applicator vocabulary is not enabled and the validator vocabulary is enabled, what is the expected behaviour of any `minContains` and `maxContains` keywords?

> I also want to push back on this terminology. Metaschemas don't disable vocabularies. They either have them or they don't. They don't have them but turn them off.

I don't know what you mean here. Doesn't the 2019-09 specification itself "have" the format vocabulary but disables it?  (The specification could be clarified here - it only mentions the boolean value false in the context of unrecognized vocabulary URIs - ref https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.8.1.2)

**Comment by handrews** [comment]
@karenetheridge garbage in garbage out.  It states clearly that using vocabularies with conflicting semantics produces undefined behavior.  That is in the spec specifically to ensure that implementations of vocabulary X do not need to know what other vocabularies are or are not present.

If X and Y define the same keyword in an incompatible way, the plugins will probably both try to do their thing and the result will be weird, quite likely just an always-fail-to-validate if they are assertions.

If X depends on a keyword from Y (which should document clearly, of course), and someone uses X with Z instead of Y, where Z happens to define that keyword in an incompatible way, again, you will get arbitrary behavior that probably makes no sense.

**The principle that conflicting keyword semantics have undefined behavior is critically important for exactly this reason.**

The spec already addresses this just fine.

**Comment by handrews** [comment]
@karenetheridge responding to edits you made while I wrote the last response:

> Perhaps the specification can simply be amended to state that minContains/maxContains behaviour is undefined if the applicator vocabulary is not also enabled.

No, that is not needed and requires implementations to check for one vocabulary from another, which is precisely what we do not want.  The burden is on the meta-schema author.  Meta-schema authoring is specifically intended as an advanced feature and therefore has a high requirement for getting things right.  This simplifies both implementation and schema authoring.

**Comment by karenetheridge** [comment]
I'm not asking about a new vocabulary defining a conflicting 'contains' keyword. I'm asking about what happens when the applicator vocabulary is not enabled, but the keyword is present. Reading your responses several times, I still don't know what to do here.

**Comment by handrews** [comment]
> I don't know what you mean here. Doesn't the 2019-09 specification itself "have" the format vocabulary but disables it?
> it only mentions the boolean value false in the context of unrecognized vocabulary URIs

"recognized" means "the code can figure out what to do with it" not "the implementer has heard of it but didn't write anything."

Nothing about the `format` vocabulary is a precedent for anything else at all ever.

**Comment by handrews** [comment]
> I'm not asking about a new vocabulary defining a conflicting 'contains' keyword. I'm asking about what happens when the applicator vocabulary is not enabled, but the 'contains' (and minContains/maxContains) keyword is present. Reading your responses several times, I still don't know what to do here.

Nothing.  You do nothing.  You implement the applicator vocabulary.  You implement the validator vocabulary.  If someone writes a meta-schema that splices in some other thing that has a `contains`, then they deserve whatever happens, which is undefined, e.g. arbitrary and unpredictable.  It would be nice to not crash, but that's about it.

If neither the applicator nor some other vocabulary containing `contains` is present, then it's a moot point, isn't it?  No `contains` means the keywords don't do anything.

Each vocabulary implementation stands alone, as does each keyword, modified only by annotation results when specified.  Yes, `minContains` and `maxContains` will be (according to my suggestion) implemented in terms of the annotation results of `contains`.  If someone creates a meta-schema that causes some other thing to supply a `contains` annotation, **it is not the responsibility of the validation vocabulary or the `contains` keyword to somehow handle it.**

Just implement each keyword and vocabulary as specified.  There is **no state among vocabularies**.

**Comment by karenetheridge** [comment]
You still haven't answered my question, but I have run out of ways to rephrase it.

Someone else can come up with a test for the test suite and argue for what they think should happen, and I'll go from there.

**Comment by handrews** [comment]
@karenetheridge what part of "don't do anything" is inadequate?  I literally cannot figure out what you think you need to do?

The implementation of `minContains` is (in terms of annotations):

* check for a `contains` annotation with the same schema location path as the current keyword
* if present, validation passes if the number of matching items is greater than or equal to the value of `minContains`, and fails otherwise
* if absent validation passes (because `minContains` has no effect)

There is nothing in here about which vocabulary provides what and that is 100% intentional.  The annotation is there or it's not, nothing else matters.

**Comment by handrews** [comment]
Also I do plan to fix `minContains` /  `maxContains` to be in terms of annotations so this needs to stay open for that PR.

**Comment by karenetheridge** [comment]
Okay, if minContains/maxContains are being rewritten to use annotations from 'contains', then that changes things a bit. However, this isn't even in the master branch of the repository, and ($deity knows I've tried) I'm not a great mindreader, so I was asking *in the context of the specification as it is written today*, because that's all I know about.

PS. this is another case of annotation dependencies across vocabularies, so the evaluation order needs to be strictly defined.

**Comment by handrews** [comment]
@karenetheridge ugh 😦  I got this confused with #924 where I [literally just explained this yesterday](https://github.com/json-schema-org/json-schema-spec/issues/924#issuecomment-693759323).  I did not realize we had two separate issues devoted to this mess.  In which case yes, let's close this in favor of #924 and avoid a repeat.  While keeping track of duplicate parallel threads is technically easier than reading minds, I'm unlikely to notice such things in the back and forth of multiple projects.

There is a concern over order of evaluation but I believe that does not require a formal vocabulary dependency.  Since we decided that annotations are always named after the keyword, and this particular dependency is on adjacent keywords only, I think there's a fairly straightforward way to handle this.  Let me think on it more, though.  I probably want to clarify the annotation and vocabulary processing model as there is plenty of non-obvious stuff here.
